### PR TITLE
PTP-838 Support Polygon

### PR DESCRIPTION
**Pull request by efecarranza**
# Changelog

Support Polygon and Binance.

Depends on: https://github.com/ConsenSys/nft-api-apps/pull/95

### CODE REVIEW

File: src/lib/Auth/availableChains.js [hunk 1]
```diff
@@ -1,18 +1,18 @@
 const Chains = {
   mainnet: 1,
   goerli: 5,
-  rinkeby: 4,
```
Comment (kalote) [inline src/lib/Auth/availableChains.js:4]: why did you remove rinkeby and ropsten?

**Review by kalote — CHANGES_REQUESTED** [review]

File: src/lib/Auth/availableChains.js (hunk 1 above)
Comment (efecarranza) [reply src/lib/Auth/availableChains.js:4]: @kalote no longer supported by Moralis

File: src/lib/Auth/availableChains.js [hunk 2]
```diff
@@ -1,18 +1,18 @@
...
-  ropsten: 3,
+  polygon: 137,
```
[the remark this answers is not in the document]
Comment (efecarranza) [reply src/lib/Auth/availableChains.js:4]: @kalote requirement was only polygon, mumbai not supported by infura.

**Comment by efecarranza** [comment]
@kalote updated

**Review by salimtb — APPROVED** [review]

File: package.json [hunk 3]
```diff
@@ -48,25 +48,25 @@
   "dependencies": {
     "axios": "^0.27.2",
     "dotenv": "^16.0.1",
-    "ethers": "^5.6.6"
+    "eslint": "^8.20.0",
```
Comment (kalote) [inline package.json:51]: eslint == dev dependencies

File: src/lib/ContractTemplates/ERC721Mintable.js [hunk 4]
```diff
@@ -27,6 +27,19 @@ export default class ERC721Mintable {
     return this.#template;
   }
 
+  /* eslint-disable class-methods-use-this */
+  addGasPriceToOptions(options, gas) {
+    const newOptions = options;
+    if (gas) {
+      if (typeof parseFloat(gas) !== 'number') {
+        throw new Error('[ERC721Mintable] Invalid value for gas provided');
```
Comment (kalote) [inline src/lib/ContractTemplates/ERC721Mintable.js:35]: put error message in errorMessages.js + use error handler for throw

File: src/lib/ContractTemplates/ERC721Mintable.js [hunk 5]
```diff
@@ -27,6 +27,19 @@ export default class ERC721Mintable {
...
+      }
+      const gasPrice = ethers.utils.parseUnits(gas, 'gwei');
```
Comment (kalote) [inline src/lib/ContractTemplates/ERC721Mintable.js:37]: if this fail, it won't be catch => add try catch block

File: src/lib/ContractTemplates/ERC721Mintable.js [hunk 6]
```diff
@@ -36,7 +49,7 @@ export default class ERC721Mintable {
    * @notice Warning: This method will consume gas (4000000 gas estimated)
    * @returns void
    */
-  async deploy({ name, symbol, contractURI }) {
+  async deploy({ name, symbol, contractURI, gas = null }) {
```
Comment (kalote) [inline src/lib/ContractTemplates/ERC721Mintable.js:52]: why not call that `options = {}` so that we have a placeholder for other options + allowing user to extend the options of the etherjs call?

File: usage.js [hunk 7]
```diff
@@ -9,14 +9,15 @@
 // Use this in "module" type nodejs projects
 import { config as loadEnv } from 'dotenv';
 import { SDK, Auth, TEMPLATES } from './index.js';
+import { ethers } from 'ethers';
```
Comment (kalote) [inline usage.js:12]: Not used => can be removed

File: usage.js [hunk 8]
```diff
@@ -9,14 +9,15 @@
...
 loadEnv();
 
 const acc = new Auth({
   privateKey: process.env.WALLET_PRIVATE_KEY,
   projectId: process.env.INFURA_PROJECT_ID,
   secretId: process.env.INFURA_PROJECT_SECRET,
-  rpcUrl: process.env.EVM_RPC_URL,
-  chainId: 4,
+  rpcUrl: 'https://polygon-mainnet.g.alchemy.com/v2/cyKpdJC2AHsuKa_NfKJ21rS-qmM2bQwP',
```
Comment (kalote) [inline usage.js:19]: we're not going to publicly publish an "infura-sdk" that uses an "alchemy" endpoint right? Is it because you were not able to get your infura account with polygon add-on enable? if that's the case, it can be solved

**Review by kalote — CHANGES_REQUESTED** [review]
few typos and small stuff to change. Implementation example  to add. Alchemy to be removed :D

**Review by danielpavelicconsensys — APPROVED** [review]
Fix Johann's comments and good to go.

File: usage.js (hunk 8 above)
Comment (efecarranza) [reply usage.js:19]: lol sorry!!! i was trying to see if alchmey threw the same error 😂  (it does!)

File: src/lib/ContractTemplates/ERC721Mintable.js (hunk 6 above)
Comment (efecarranza) [reply src/lib/ContractTemplates/ERC721Mintable.js:52]: gets a bit convoluted with our gas limit that we have to set 😬 
can be updated in a follow up PR if we think it works better

File: src/lib/ContractTemplates/ERC721Mintable.js [hunk 9]
```diff
@@ -27,6 +27,31 @@ export default class ERC721Mintable {
     return this.#template;
   }
 
+  /* eslint-disable class-methods-use-this */
+  addGasPriceToOptions(options, gas) {
```
Comment (kalote) [inline src/lib/ContractTemplates/ERC721Mintable.js:31]: Should we make it private?
- yes => make it private
- no => add JSDoc

**Review by kalote — CHANGES_REQUESTED** [review]
small visibility correction and we're good

**Review by kalote — APPROVED** [review]
